### PR TITLE
PS-269: Fix MTR --do-test=percona_heap (8.0)

### PR DESCRIPTION
**Pull request by inikep**
1. Fix MTR tests with `--do-test=percona_heap` and re-record them
2. Fix ASAN warnings: `load/store of misaligned address` in hp_create.cc and hp_dspace.cc

### CODE REVIEW

File: mysql-test/r/percona_heap_blob.result [hunk 1]
```diff
@@ -460,67 +457,66 @@ Field	Type	Collation	Null	Key	Default	Extra	Privileges	Comment
 l	longblob	NULL	YES		NULL		#	
 drop table t1;
 create table t1 (id integer primary key auto_increment, txt text not null);
-insert into t1 (txt) values ('Chevy ');
+insert into t1 (txt) values ('Chevy');
```
Comment (laurynas-biveinis) [inline mysql-test/r/percona_heap_blob.result]: Why?

File: mysql-test/r/percona_heap_blob.result [hunk 2]
```diff
@@ -460,67 +457,66 @@ Field	Type	Collation	Null	Key	Default	Extra	Privileges	Comment
...
-1	Chevy 
+1	Chevy
 2	Ford
 drop table t1;
 create table t1 (id integer primary key auto_increment, txt text);
 insert into t1 (txt) values ('Chevy'), ('Chevy '), (NULL);
 select * from t1 where txt='Chevy' or txt is NULL;
 id	txt
 1	Chevy
-2	Chevy
```
Comment (laurynas-biveinis) [inline mysql-test/r/percona_heap_blob.result]: Where did "Chevy " go?

File: mysql-test/r/percona_heap_blob.result [hunk 3]
```diff
@@ -798,38 +786,44 @@ ERROR 42000: Display width out of range for column 'a' (max = 4294967295)
 CREATE TABLE b15776 (a year(999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999));
 ERROR HY000: Supports only YEAR or YEAR(4) column.
 CREATE TABLE b15776 select cast(null as char(4294967295));
+Warnings:
+Warning	1301	Result of cast_as_char() was larger than max_allowed_packet (67108864) - truncated
```
Comment (laurynas-biveinis) [inline mysql-test/r/percona_heap_blob.result:889]: TODO bump max_allowed_packet?

File: mysql-test/r/percona_heap_var.result [hunk 4]
```diff
@@ -76,7 +75,7 @@ Name	t1
 Engine	MEMORY
 Version	10
 Row_format	Dynamic
-Rows	7
+Rows	0
```
Comment (laurynas-biveinis) [inline mysql-test/r/percona_heap_var.result:78]: We discussed this on Slack, that it is a valid value, but why is this cached at 0 now?

File: mysql-test/t/percona_heap_blob.test [hunk 5]
```diff
@@ -24,7 +24,7 @@ CREATE TABLE t1 (a blob, b text, c blob(250), d text(70000), e text(70000000));
 show columns from t1;
 # PS doesn't give errors on prepare yet
 CREATE TABLE t2 (a char(255), b varbinary(70000), c varchar(70000000));
-CREATE TABLE t4 (c varchar(65530) character set utf8 not null);
+CREATE TABLE t4 (c varchar(65530) character set utf8mb4 not null);
```
Comment (laurynas-biveinis) [inline mysql-test/t/percona_heap_blob.test]: case mismatch on the whole line

File: mysql-test/t/percona_heap_blob.test [hunk 6]
```diff
@@ -487,21 +487,21 @@ CREATE TABLE b15776 (a year(9999999999999999999999999999999999999999999999999999
 CREATE TABLE b15776 select cast(null as char(4294967295));
 show columns from b15776;
 drop table b15776;
-CREATE TABLE b15776 select cast(null as nchar(4294967295));
+CREATE TABLE b15776 select cast(null as char(4294967295) CHARACTER SET UTF8MB4);
```
Comment (laurynas-biveinis) [inline mysql-test/t/percona_heap_blob.test]: case mismatch for the whole line

File: mysql-test/t/percona_heap_blob.test [hunk 7]
```diff
@@ -487,21 +487,21 @@ CREATE TABLE b15776 (a year(9999999999999999999999999999999999999999999999999999
...
 show columns from b15776;
 drop table b15776;
 CREATE TABLE b15776 select cast(null as binary(4294967295));
 show columns from b15776;
 drop table b15776;
 
 explain select cast(1 as char(4294967295));
-explain select cast(1 as nchar(4294967295));
+explain select cast(1 as char(4294967295) CHARACTER SET UTF8MB4);
```
Comment (laurynas-biveinis) [inline mysql-test/t/percona_heap_blob.test]: Case mismatch for the whole line

File: mysql-test/t/percona_heap_blob.test [hunk 8]
```diff
@@ -487,21 +487,21 @@ CREATE TABLE b15776 (a year(9999999999999999999999999999999999999999999999999999
...
 explain select cast(1 as binary(4294967295));
 
 --error ER_TOO_BIG_DISPLAYWIDTH
 explain select cast(1 as char(4294967296));
 --error ER_TOO_BIG_DISPLAYWIDTH
-explain select cast(1 as nchar(4294967296));
+explain select cast(1 as char(4294967296) CHARACTER SET UTF8MB4);
```
Comment (laurynas-biveinis) [inline mysql-test/t/percona_heap_blob.test]: Case mismatch for the whole line

File: mysql-test/t/percona_heap_blob.test [hunk 9]
```diff
@@ -525,11 +525,11 @@ explain select convert(1, char(4294967295));
 explain select convert(1, char(4294967296));
 --error ER_TOO_BIG_DISPLAYWIDTH
 explain select convert(1, char(999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999));
-explain select convert(1, nchar(4294967295));
+explain select convert(1, char(4294967295) CHARACTER SET UTF8MB4);
```
Comment (laurynas-biveinis) [inline mysql-test/t/percona_heap_blob.test]: Case mismatch for the whole line

File: mysql-test/t/percona_heap_blob.test [hunk 10]
```diff
@@ -525,11 +525,11 @@ explain select convert(1, char(4294967295));
...
 --error ER_TOO_BIG_DISPLAYWIDTH
-explain select convert(1, nchar(4294967296));
+explain select convert(1, char(4294967296) CHARACTER SET UTF8MB4);
```
Comment (laurynas-biveinis) [inline mysql-test/t/percona_heap_blob.test]: Case mismatch for the whole line

File: mysql-test/t/percona_heap_blob.test [hunk 11]
```diff
@@ -525,11 +525,11 @@ explain select convert(1, char(4294967295));
...
 --error ER_TOO_BIG_DISPLAYWIDTH
-explain select convert(1, nchar(999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999));
+explain select convert(1, char(999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999) CHARACTER SET UTF8MB4);
```
Comment (laurynas-biveinis) [inline mysql-test/t/percona_heap_blob.test]: Case mismatch for the whole line

File: sql/item.cc [hunk 12]
```diff
@@ -2302,7 +2302,8 @@ Item_field::Item_field(Field *f)
       no_const_subst(false),
       have_privileges(0),
       any_privileges(false) {
-  if (f->table->pos_in_table_list != NULL)
+  if (f->table->pos_in_table_list != NULL &&
+      f->table->pos_in_table_list->select_lex != NULL)
```
Comment (laurynas-biveinis) [inline sql/item.cc]: nullptr

Comment (laurynas-biveinis) [inline sql/item.cc]: Split out to a separate commit

File: storage/heap/hp_dspace.cc [hunk 13]
```diff
@@ -234,8 +235,10 @@ static uchar *hp_allocate_variable_chunkset(HP_DATASPACE *info,
       if (last_existing_chunk) {
         /* Truncate whatever was added at the end of the existing chunkset */
         prev_chunk = last_existing_chunk;
-        curr_chunk = *((uchar **)(prev_chunk + info->offset_link));
-        *((uchar **)(prev_chunk + info->offset_link)) = nullptr;
+        memcpy(&curr_chunk, prev_chunk + info->offset_link, sizeof(uchar *));
+        memset(reinterpret_cast<uchar **>(prev_chunk + info->offset_link), 0,
```
Comment (laurynas-biveinis) [inline storage/heap/hp_dspace.cc:239]: Very strictly speaking the nullptr representation is not required to be all zeros - but fixing this - say, on read checking if all bytes are 0, if yes, setting to nullptr - seems waste of effort here. A better fix would be to have explicit chunk pointers instead, but this is a bigger refactoring than is currently justifiable

File: storage/heap/hp_record.cc [hunk 14]
```diff
@@ -267,17 +268,19 @@ void hp_copy_record_data_to_chunkset(const HP_SHARE &info, const uchar *record,
   src_offset.
 */
 #if !defined(DBUG_OFF) && defined(EXTRA_HEAP_DEBUG)
-#define SWITCH_TO_NEXT_CHUNK_FOR_READ(share, curr_chunk, src_offset)         \
-  {                                                                          \
-    curr_chunk = *((uchar **)(curr_chunk + share->recordspace.offset_link)); \
-    src_offset = 0;                                                          \
-    dump_chunk(share, curr_chunk);                                           \
+#define SWITCH_TO_NEXT_CHUNK_FOR_READ(share, curr_chunk, src_offset) \
+  {                                                                  \
```
Comment (laurynas-biveinis) [inline storage/heap/hp_record.cc]: Since you are touching this anyway, maybe convert to an inline function with inout reference args? Or at least do the proper do { ... } while(0) compound-statement-in-a-macro thing

**Review by laurynas-biveinis — CHANGES_REQUESTED** [review]

File: mysql-test/r/percona_heap_var.result (hunk 4 above)
Comment (inikep) [reply mysql-test/r/percona_heap_var.result:78]: Because it's before 1st INSERT. It seems to always cache a result of first `SHOW TABLE STATUS LIKE "t1";` call.

File: mysql-test/r/percona_heap_blob.result (hunk 3 above)
Comment (inikep) [reply mysql-test/r/percona_heap_blob.result:889]: This is supposed to give a warning. A comment says:
```
## Do not select, too much memory needed.
```

File: storage/heap/hp_record.cc [hunk 15]
```diff
@@ -266,20 +267,16 @@ void hp_copy_record_data_to_chunkset(const HP_SHARE &info, const uchar *record,
   Macro to switch curr_chunk to the next chunk in the chunkset and reset
```
Comment (laurynas-biveinis) [inline storage/heap/hp_record.cc]: Not a "Macro" now

File: storage/heap/hp_record.cc [hunk 16]
```diff
@@ -266,20 +267,16 @@ void hp_copy_record_data_to_chunkset(const HP_SHARE &info, const uchar *record,
...
   src_offset.
 */
+static inline void switch_to_next_chunk_for_read(const HP_SHARE *const share,
```
Comment (laurynas-biveinis) [inline storage/heap/hp_record.cc]: Can share arg be made const ref?

**Review by laurynas-biveinis — CHANGES_REQUESTED** [review]

File: storage/heap/hp_record.cc [hunk 17]
```diff
@@ -263,23 +264,19 @@ void hp_copy_record_data_to_chunkset(const HP_SHARE &info, const uchar *record,
 }
 
 /*
-  Macro to switch curr_chunk to the next chunk in the chunkset and reset
+  Function to switch curr_chunk to the next chunk in the chunkset and reset
   src_offset.
 */
+static inline void switch_to_next_chunk_for_read(const HP_SHARE *const &share,
```
Comment (laurynas-biveinis) [inline storage/heap/hp_record.cc]: const HP_SHARE &share ?

**Review by laurynas-biveinis — CHANGES_REQUESTED** [review]

**Review by laurynas-biveinis — APPROVED** [review]
